pkg/monitortests/kubeapiserver/crdversionchecker/monitortest.go

```diff
@@ -0,0 +1,486 @@
+		}
+	}
+	sort.Strings(removedVersions)
+
+	// Check if storage version changed
+	storageChanged := beforeStorage != afterStorage
+
+	// Only return a change summary if something actually changed
+	if len(addedVersions) == 0 && len(removedVersions) == 0 && !storageChanged {
+		return nil
+	}
+
+	return &CRDChangeSummary{
+		Name:            name,
+		AddedVersions:   addedVersions,
+		RemovedVersions: removedVersions,
+		StorageChanged:  storageChanged,
+		OldStorage:      beforeStorage,
+		NewStorage:      afterStorage,
+	}
+}
+
+// =============================================================================
+// Validation Checks
+// =============================================================================
+
+// checkNewVersionsNotStoredImmediately verifies that newly added API versions
+// are not immediately set as the storage version.
+//
+// Rationale: When a new API version is introduced during an upgrade, existing
+// data in etcd is still stored in the old format. Setting the new version as
+// the storage version immediately would require a migration. The safe approach
+// is to serve both versions but keep the old version as storage until all
+// existing objects have been migrated.
+//
+// This check FAILS if:
+// - A CRD has a new version added
+// - The new version is marked as the storage version
+// - The old version is no longer the storage version
+//
+// This check PASSES if:
+// - New versions are added but the old storage version is preserved
+// - Both old and new versions are served
+func (c *crdVersionChecker) checkNewVersionsNotStoredImmediately() []*junitapi.JUnitTestCase {
+	const testName = "[sig-api-machinery] CRDs with new API versions should not change storage version immediately"
+
+	if c.beforeSnapshot == nil || c.afterSnapshot == nil {
+		return []*junitapi.JUnitTestCase{{
+			Name: testName,
+			SkipMessage: &junitapi.SkipMessage{
+				Message: "Missing CRD snapshots, cannot perform check",
+			},
+		}}
+	}
+
+	var failures []string
+
+	for crdName, beforeCRD := range c.beforeSnapshot.CRDs {
+		afterCRD, exists := c.afterSnapshot.CRDs[crdName]
+		if !exists {
+			// CRD was removed, not relevant for this check
+			continue
+		}
+
+		// Identify the storage version before and after
+		beforeStorage := getStorageVersion(beforeCRD)
+		afterStorage := getStorageVersion(afterCRD)
+
+		// Identify newly added versions
+		beforeVersionSet := make(map[string]bool)
+		for _, v := range beforeCRD.Versions {
+			beforeVersionSet[v.Name] = true
+		}
+
+		newVersions := []string{}
+		for _, v := range afterCRD.Versions {
+			if !beforeVersionSet[v.Name] {
+				newVersions = append(newVersions, v.Name)
+			}
+		}
+
+		// If no new versions were added, nothing to check
+		if len(newVersions) == 0 {
+			continue
+		}
+
+		// Check if a new version became the storage version
+		for _, newVersion := range newVersions {
+			if afterStorage == newVersion && beforeStorage != newVersion {
```

Comment:
The first part of this if statement is sufficient. If it's a new version and the storage matches the new version that's added, we can fail

Don't need to check the second half (how could the previous storage be a new version? I don't think the second check can ever be false)